### DISCUSSION TITLE
How to import a monorepo workspace package AND its dependencies

### DISCUSSION THREAD

**Discussion by 58bits** [pnpm/pnpm#7113] [Q&A] [answered]
We have a typical PNPM workspace monorepo setup with our `pnpm-workspace.yaml`  like this...

```
packages:
  - 'packages/**'
  - 'apps/**'
```

One of our packages is a React UI kit, based on Radix UI, and compiled as an esm library using [tsup](https://github.com/egoist/tsup).

We'd like to import the uikit package into app01 along with the uikit dependencies - such as ...

`@ourproject/app01` - package.json
```
{
  "name": "@ourproject/app01",
  "version": "1.0.0",
  "type": "module",
  "dependencies": {
     ...
      "@ourproject/uikit": "workspace:*",
     ...
  }
}
```

`@ourproject/uikit`- package.json
```
{
  "name": "@ourproject/uikit",
  "version": "1.0.0",
  "type": "module",
  "dependencies": {
     ...
    "@radix-ui/react-accordion": "^1.1.2",
    "@radix-ui/react-avatar": "^1.0.3",
    "@radix-ui/react-checkbox": "^1.0.4",
    "@radix-ui/react-dialog": "^1.0.4",
     ...
  }
}
```

But when we do this - we _only_ receive our compiled code, and NOT any of @ourproject/uikit 's dependencies (the radix-ui dependencies are missing from @outproject/app01). When trying to use our uikit inside @ourproject/app01 it will complain about missing radix-ui dependencies, and so we then have to manually add all the radix-ui deps directly the package.json file for app01.

I've read the pnpm docs and have tried to make sense of hoisting, peer dependencies and other, but can't seem to figure out a way for this to 'just work' (as it would for example, with an npm package that installs along with its dependencies).

Probably missing something simple.  Thoughts or suggestions greatly appreciated.

**Comment by stevenpetryk** [comment]
Are you sure that you have symlinks supported in whatever is compiling your apps?

What should be happening is that the bundler would resolve uikit's dependencies from packages/uikit/node_modules. But if it's not following the symlink into that directory, it won't find the dependencies. 

It would be useful to provide the exact error from your development server, as it may show the attempted resolution paths.

**Reply by 58bits** [reply]
Hi @stevenpetryk - thanks for taking the time to reply.  

Our 'app01' is a [Remix Run](https://remix.run/) application and these are the build errors we receive...

```
: ✘ [ERROR] Could not resolve "@radix-ui/react-dropdown-menu"
remix:build:
remix:build:  app/ui/layouts/app-menu.tsx:1:39:
remix:build:    1 │ import * as DropdownMenuPrimitive from '@radix-ui/react-dropdown-menu'
remix:build:         ╵                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can mark the path "@radix-ui/react-icons" as external to exclude it from the bundle, which will remove this error.
```

**Reply by stevenpetryk** [reply, marked as answer]
It looks like that error originated from code in app/. As in, you are importing Radix UI from your app directly. But radix is a dependency of UIKit, not your app. pnpm doesn't let you access transitive dependencies like this, as a feature.

**Reply by 58bits** [reply]
Hi again @stevenpetryk  - zomg - you're right. I just searched for direct imports of radix components and discovered some older theme and menu components that were still importing radix-ui directly (face palm). [And yes - I'm aware of the pnpm feature that restricts transitive imports]. Thanks again for taking the time. It's solved.
